Simplify internal/fs

greatroar opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

It introduces various simplifications to the internal/fs package. In particular, the Write method on fs.File is gone, fs has been decoupled from testing, and fixpath has been simplified to use the [stdlib's handling of long path names](https://golang.org/doc/go1.8#os) on Windows.


Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

No.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

Commit:
Remove io.Writer from fs.File

It was only used in a single test, which now uses plain *os.File instead.

Commit:
Move internal/fs.TestChdir to internal/test.Chdir

Commit:
Remove unused fs.Rename

Commit:
Fix and simplify fs.Reader

fakeDir.{Readdir,Readdirnames} weren't handling the case n == 0
correctly. fakeFileInfo.sys is always nil, so omit the field.

Commit:
Remove OS-specific versions of fs.MkdirAll

Go has supported Windows paths correctly since 1.11, see
https://github.com/golang/go/issues/10900 and the commit referenced
there.

MichaelEischer commented on internal/textfile/read_test.go:
```diff
@@ -60,10 +68,15 @@ func TestRead(t *testing.T) {
-			f, cleanup := writeTempfile(t, test.data)
-			defer cleanup()
+			tempname := writeTempfile(t, test.data)
+			defer func() {
```
I'd like to keep the cleanup function as a return value of `writeTempfile`, so that a user just has to call `defer cleanup()` as before

greatroar replied:
Done.

MichaelEischer reviewed (commented):
Please split the `fixpath` changes into a separate PR. The other changes are fine.

As reported by @JaCoB1123 the go standard library only prepends `\\?\` to long filepaths. However, that way short paths are still normalized which breaks backing up files with reserved filenames such as `test.` or `con.m`.

I also think that the fix to properly handle special filenames on windows has to be more extensive: With these changes the backup command will fail for the special files when using a relative path for backup  (we could maybe normalize the path at https://github.com/restic/restic/blob/34a3adfd8d2b84a5cf121e94174ff2dc94dd4a23/internal/archiver/archiver.go#L550 ). And the restorer in turn fails to restore special filenames (https://github.com/restic/restic/blob/34a3adfd8d2b84a5cf121e94174ff2dc94dd4a23/internal/restorer/restorer.go#L194).
May idea here would be to fix the file paths to include `\\?\` as soon as possible, as afterwards these paths will just behave properly.

Commit:
Remove stray Printf from internal/fs

greatroar commented:
Force-pushed a new version with the Windows path handling removed, except for the MkdirAll case. Also removed a stray Printf from fs_reader.go. Sorry for the wait.

Commit:
Simplify os.ModeType|os.ModeCharDevice => os.ModeType

Since Go 1.12, ModeCharDevice is included in ModeType:
golang/go@a2a3dd00c934fa15ad880ee5fe1f64308cbc73a7

greatroar commented:
Also removed a Go < 1.12 workaround.

MichaelEischer commented on internal/textfile/read_test.go:
```diff
@@ -3,25 +3,40 @@ package textfile
+	}()
 
-	err = f.Close()
+	_, err = f.Write(data)
```
Shouldn't the test also delete the tempfile if the write fails? Maybe add a add a `shouldDelete` flag and check that in the deferred function?

greatroar replied:
Good catch. I found a different way of doing this.

MichaelEischer reviewed (commented):
> Sorry for the wait.

No problem. There are plenty other PR to keep me busy ;-)

The PR is basically ready for merging, I've just noticed a small, already existing bug in the textfile read test, which would be good to fix since the PR already modifies the code.

I'm amazed by how many workarounds for old go version you uncover all the time :-) , thanks for that. The old `MkdirAll` didn't seem to call `fixpath`, so the restorer might have actually failed to restore directories with certain special names.

MichaelEischer reviewed (approved):
LGTM. Thanks for the cleanup :-)
